lib/contract.ts

```diff
@@ -1032,16 +1032,102 @@ export default class Contract {
 			[] as Contract[],
 		);
 	}
+
+	private isRequirementSatisfied(
+		requirement: Contract,
+		options: { types?: Set<string> } = {},
+	): boolean {
+		// Utilities
+		const shouldEvaluateType = (type: string) =>
+			options.types ? options.types.has(type) : true;
+
+		/**
+		 * @summary Check if a matcher is satisfied
+		 * @function
+		 * @public
+		 *
+		 * @param {Object} matcher - matcher contract
+		 * @returns {Boolean} whether the matcher is satisfied
+		 *
+		 * @example
+		 * const matcher = Contract.createMatcher({
+		 *   type: 'sw.os',
+		 *   slug: 'debian'
+		 * })
+		 *
+		 * if (hasMatch(matcher)) {
+		 *   console.log('This matcher is satisfied!')
+		 * }
+		 */
+		const hasMatch = (matcher: Contract): boolean => {
+			// TODO: Write a function similar to findContracts
+			// that stops as soon as it finds one match
+			return (
+				this.findChildren(matcher).length > 0 ||
+				this.findChildrenWithCapabilities(matcher).length > 0
+			);
+		};
+
+		if (requirement.raw.operation === 'or') {
+			// (3.1) Note that we should only consider disjuncts
+			// of types we are allowed to check. We can make
+			// such transformation here, so we can then consider
+			// the disjunction as fulfilled if there are no
+			// remaining disjuncts.
+			const disjuncts = filter(requirement.raw.data.getAll(), (disjunct) => {
+				return shouldEvaluateType(disjunct.raw.data.type);
+			});
+			// (3.2) An empty disjuction means that this particular
+			// requirement is fulfilled, so we can carry on.
+			// A disjunction naturally contains a list of further
+			// requirements we need to check for. If at least one
+			// of the members is fulfilled, we can proceed with
+			// next requirement.
+			if (disjuncts.length === 0 || some(disjuncts, hasMatch)) {
+				return true;
+			}
```

Comment:
I'm less confident on this combination but it should work (I may have done the conversion incorrectly, but the conversion into a single `every` should be possible)
```suggestion
			// (3.2) An empty disjuction means that this particular
			// requirement is fulfilled, so we can carry on.
			// A disjunction naturally contains a list of further
			// requirements we need to check for. If at least one
			// of the members is fulfilled, we can proceed with
			// next requirement.
			if (every(requirement.raw.data.getAll(), (disjunct) => {
				return !shouldEvaluateType(disjunct.raw.data.type) || !hasMatch(disjunct)
			})) {
				return true;
			}
```

Reply:
This doesn't work unfortunately, in order for the requirement to be accepted we need either: 

- there are no requirements to evaluate
- if there are requirements to evaluate, there is at least one match


The closest thing would be 

```ts
if (
	every(requirement.raw.data.getAll(), (disjunct) => {
		return (
			!shouldEvaluateType(disjunct.raw.data.type) || hasMatch(disjunct)
		);
	})
) {
	return true;
}
```

however this is not the same, because if `shouldEvaluateType(disjunct.raw.data.type)` is always true, this is equivalent to writing

```ts
if (
	every(requirement.raw.data.getAll(), (disjunct) => {
		return hasMatch(disjunct);
	})
) {
	return true;
}
```

which is the opposite of what we want

I'm not sure there is an equivalent using a single `some/every`

Reply:
Yeah, I think the problem is we need `every(disjuncts, !shouldEvaluateType) || some(disjuncts, shouldEvaluateType && hasMatch)` and those can't be combined so what we currently have is the best we can manage with functional. However it can be done via a for-loop with `return` to short-circuit the `some` case but keeping the `every` case requiring an entire loop:
```suggestion
			// (3.2) An empty disjuction means that this particular
			// requirement is fulfilled, so we can carry on.
			// A disjunction naturally contains a list of further
			// requirements we need to check for. If at least one
			// of the members is fulfilled, we can proceed with
			// next requirement.
			let noneToEvaluate = true;
			for (const disjunct of (requirement.raw.data.getAll()) {
				if (shouldEvaluateType(disjunct.raw.data.type)) {
					noneToEvaluate = false
					if (hasMatch(disjunct)) {
						return true;
					}
				}
			}
			return noneToEvaluate;
```

Fwiw I don't think this is worth it unless it's a performance bottleneck but since I dug into how to do it I thought I'd share

Reply:
It does sacrifice a bit of readability, I suspect at some point we'll need to work on the efficiency of some of the operations at which point we can review